### PR TITLE
ci: build and publish image

### PR DESCRIPTION
**Pull request by tisonkun**
Signed-off-by: tison <wander4096@gmail.com>

Here is a draft PR for building and publish container image of `engula`. Given that binary in main branch is unstable, I make this draft based on v0.2.0 and hope that we can migrate to the main branch later. This is a preview and I'm going to make use of this image to setup a docker-compose stack for running e2e tests.

Try it out:

```
docker pull ghcr.io/tisonkun/engula:latest
docker run -it ghcr.io/tisonkun/engula:latest # <args>
```

Pending issues:

* [x] `docker/login-action` should login by who has the permission to publish to the repository packages, see also [how rust-lang/docker-rust-nightly did it](https://github.com/rust-lang/docker-rust-nightly/blob/312f155d045659692f36b86170a61a7e544320f0/.github/workflows/build.yml#L45-L55).
    Non issue as I notice that `github.actor` of pushing by merging is the merge actor (i.e., a comitter).
* [x] Reconcile nightly publishes and release publishes.
* Multi-platform support and slim the image are possible, but I tend to leave it as improvements.

Reference:

* #226 
* #170

### CODE REVIEW

**Comment by zojw** [comment]
how about helping add `target` folder to `.dockerignore` too~

developer usually `cargo build` in local and do some local test then `docker build` do more test in k8s env.

target folder is big and waste time in `COPY .  /build/` in my local laptop, it's 7.5G and 30s

**Comment by tisonkun** [comment]
@zojw good point. Let me push a follow up for it.

File: docker-compose.yml [hunk 1]
```diff
@@ -18,18 +18,20 @@ services:
...
     command: storage run 0.0.0.0:10002 --mem
   kernel:
     image: ghcr.io/tisonkun/engula:nightly
-    ports:
-      - "10003:10003"
+    network_mode: "service:journal"
     depends_on:
       - journal
       - storage
-    command: kernel run 0.0.0.0:10003 --journal journal:10001 --storage storage:10002 --mem
```
Comment (tisonkun) [inline docker-compose.yml]: With an offline discussion with @zheaoli I make this network hack. However, a clean solution should be separating runtime network config and client side network config - they have different resolutions.

@zheaoli you can start a discussion on the solution and refer to this commit as an evidence.

Comment (Zheaoli) [reply docker-compose.yml]: > you can start a discussion on the solution and refer to this commit as an evidence.

Sure I will create a new discussion ASAP tomorrow.

Comment (tisonkun) [reply docker-compose.yml]: A bit details on this commit:

1. containers in docker-compose services have dedicated namespace ([ref](https://docs.docker.com/compose/networking/))
2. the namespace (alias) is container name
3. Engula's Kernel client asks the server for getting the address of journal & storage
4. If (3) happens on host, the address of journal is `journal:10001`, but in host there is not such an alias, and thus transport error occurs.

**Comment by tisonkun** [comment]
@huachaohuang docker image building and publishing logic is ready. The blocker is that binary is opt-out on main branch. You can ping in this thread when the binary compiled again.

**Comment by tisonkun** [comment]
Closed as rebasing on main branch later.

This should close #268, close #226.

**Pull request closed without merging**